### PR TITLE
Feature/delete rename todo

### PR DESCRIPTION
**Pull request by AKallala**
This is PR implements the deleting and renaming of todo in the todo list, it:
- Adds a `DropDownMenu` to the `TodoListScreen` that is displayed when the user clicks on the three dots button of a todo item and offers two possible actions to perform on a todo, namely delete and rename
- Adds the `renameTodo` function to the `TodoListViewModel`
- Adds a dialog with a text field that is displayed to the user and through which he can rename the todo
- Refactors the `TodoItem` composable so that it can be reused reused in the pomodoro screen 

Capture d’écran 2024-11-14 à 02 27 23
Capture d’écran 2024-11-14 à 02 27 44

### Suggestions for future improvements:
- Make the `Text` composable turn into a `TextField` for renaming instead of displaying a dialog, this way the user can rename his todos in an easier and faster way

### Checklist:
- [x] Update the `TodoListScreen`
- [x] Add `renameTodo` to `TodoListViewModel` 
- [x] Add a test for `renameTodo` in `TodoListViewModelTest`
- [x] Add tests to `TodoListScreenTest` to test the new elements of `TodoListScreen`
- [x] Document code
- [x] Format code with Ktfmt

### CODE REVIEW

File: app/src/main/java/com/github/se/eduverse/ui/todo/TodoListScreen.kt [hunk 1]
```diff
@@ -56,14 +61,37 @@ fun TodoListScreen(navigationActions: NavigationActions, todoListViewModel: Todo
...
+                          { todoListViewModel.deleteTodo(todo.uid) },
+                          { showRenameDialog = true })
+                    }
+                  }
+              if (showRenameDialog) {
+                RenameTodoDialog(
+                    todo.name,
+                    { newName -> todoListViewModel.renameTodo(todo, newName) },
+                    { showRenameDialog = false })
+              }
```
Comment (MathieuxC) [inline app/src/main/java/com/github/se/eduverse/ui/todo/TodoListScreen.kt:94]: lgtm

File: app/src/main/java/com/github/se/eduverse/ui/todo/TodoListScreen.kt [hunk 2]
```diff
@@ -112,11 +152,8 @@ fun TodoItem(todo: Todo, onUndo: (Todo) -> Unit, onDone: (Todo) -> Unit) {
...
               textDecoration = if (completed) TextDecoration.LineThrough else TextDecoration.None,
               color = if (completed) Color.Gray else Color.Unspecified)
-          IconButton(
-              onClick = {},
-              modifier = Modifier.testTag("todoOptionsButton_${todo.uid}").padding(8.dp)) {
-                Icon(Icons.Default.MoreVert, contentDescription = "Todo Options")
-              }
+
+          rightMostButton()
         }
```
Comment (MathieuxC) [inline app/src/main/java/com/github/se/eduverse/ui/todo/TodoListScreen.kt:157]: A lot of values are hard-coded, be careful that it works also on phone with a different size than yours

File: app/src/main/java/com/github/se/eduverse/ui/todo/TodoListScreen.kt [hunk 3]
```diff
@@ -160,3 +198,91 @@ fun AddTodoEntry(onAdd: (String) -> Unit) {
...
+
+        DropdownMenuItem(
+            text = { Text("Rename") },
+            onClick = {
+              onRename()
+              onDismiss()
+            },
+            modifier = Modifier.testTag("renameTodoButton_${todoId}"))
+      }
+}
```
Comment (MathieuxC) [inline app/src/main/java/com/github/se/eduverse/ui/todo/TodoListScreen.kt:239]: lgtm

File: app/src/main/java/com/github/se/eduverse/ui/todo/TodoListScreen.kt [hunk 4]
```diff
@@ -160,3 +198,91 @@ fun AddTodoEntry(onAdd: (String) -> Unit) {
...
+      dismissButton = {
+        TextButton(
+            modifier = Modifier.testTag("renameTodoDismissButton"),
+            onClick = { onDismiss() },
+            colors =
+                ButtonColors(
+                    Color.Transparent, Color(0xFF217384), Color.Transparent, Color.LightGray)) {
+              Text("Cancel")
+            }
+      })
```
Comment (MathieuxC) [inline app/src/main/java/com/github/se/eduverse/ui/todo/TodoListScreen.kt:287]: lgtm

File: app/src/main/java/com/github/se/eduverse/ui/todo/TodoListScreen.kt [hunk 5]
```diff
@@ -160,3 +198,91 @@ fun AddTodoEntry(onAdd: (String) -> Unit) {
...
+/**
+ * Composable that represents a dialog to rename a todo
+ *
+ * @param todoName the current name of the todo
+ * @param onRename code executed when the rename button is clicked
+ * @param onDismiss code executed when the dialog is dismissed
+ */
+@Composable
+fun RenameTodoDialog(todoName: String, onRename: (String) -> Unit, onDismiss: () -> Unit) {
+  var newName by remember { mutableStateOf(TextFieldValue(todoName, TextRange(todoName.length))) }
```
Comment (MathieuxC) [inline app/src/main/java/com/github/se/eduverse/ui/todo/TodoListScreen.kt:250]: I feel like it is better to hold data inside a state flow, but if it works it's ok

File: app/src/main/java/com/github/se/eduverse/viewmodel/TodoListViewModel.kt [hunk 6]
```diff
@@ -105,4 +105,15 @@ class TodoListViewModel(private val todoRepository: TodoRepository) : ViewModel(
...
+  fun renameTodo(todo: Todo, newName: String) {
+    val newTodo = todo.copy(name = newName)
+    todoRepository.updateTodo(
+        todo = newTodo,
+        onSuccess = {
+          getActualTodos()
+          getDoneTodos()
+        },
+        onFailure = { exception -> Log.e("renameTodo", "Error renaming todo", exception) })
+  }
```
Comment (MathieuxC) [inline app/src/main/java/com/github/se/eduverse/viewmodel/TodoListViewModel.kt:118]: lgtm

**Review by MathieuxC — APPROVED** [review]
lgtm

File: app/src/main/java/com/github/se/eduverse/ui/todo/TodoListScreen.kt (hunk 5 above)
Comment (AKallala) [reply app/src/main/java/com/github/se/eduverse/ui/todo/TodoListScreen.kt:250]: TextFieldValue is not a function, it's a class that holds information about the value in the text field, so here the mutable state flow holds an instance of the class

**Pull request merged**